### PR TITLE
boards: add Adafruit Feather nrf52840 Express UF2

### PR DESCRIPTION
**Pull request by jacobw**
Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.

This PR includes converting the boards to variants to be consistent with hwmv2 and remove unnecessary unique boards.

### CODE REVIEW

**Comment by jadejr** [comment]
The xiao_ble board  defaults to uf2.  They chose uf2 by default since that's how the hardware ships. I'm guessing the feather defaults to utf2, but default is reversed. It'd be nice if there was a common approach here across all zephyr boards.

**Comment by jacobw** [comment]
> The xiao_ble board defaults to uf2.

the xiao_ble doesn’t have an SWD connector does it? I think the current approach is default to SWD if it has a connector, and UF2 if not (and pre flashed with the boot loader).

the Sense has SWD headers but no connector, so only has UF2 support.

**Comment by jadejr** [comment]
> > The xiao_ble board defaults to uf2.
> 
> the xiao_ble doesn’t have an SWD connector does it? I think the current approach is default to SWD if it has a connector, and UF2 if not (and pre flashed with the boot loader).
> 
Both versions have swd pins on the bottom which you can use with their expansion board or pogo pins.  I'm using their expansion board plus a nordic dev kit to flash the non-Sense version. You can see them right here 
image

**Comment by jacobw** [comment]
> > > The xiao_ble board defaults to uf2.
> > 
> > 
> > the xiao_ble doesn’t have an SWD connector does it? I think the current approach is default to SWD if it has a connector, and UF2 if not (and pre flashed with the boot loader).
> 
> Both versions have swd pins

That's why I suggested connector. Sense has pads but not headers, which would be even harder to use. 

Overall, I couldn't find an easy way to support both without having multiple boards. I was hoping to do something cleaner with the variants in hwmv2, but that didn't seem possible. Not sure there is a good answer to this.

**Comment by jadejr** [comment]
I'm just suggesting whatever the answer is should be applied to all boards that default to utf2.

**Comment by jacobw** [comment]
> I'm just suggesting whatever the answer is (good or not) should be applied to all boards that default to utf2.

I've been wondering what the options are for dealing with this in a consistent way.

I thought  UF2 required specific board level definitions to set the Kconfig options, but looking xiao_ble_defconfig perhaps that can all be managed as board variants. In that case, perhaps an approach is to always default to SWD, and have a UF2 variant for any board that needs/wants it.

That would mean creating new non-UF2 definitions and that would require maintenance so perhaps that is too much effort in the long term. Or does that just mean people will be creating custom boards out of tree?

**Comment by jacobw** [comment]
PR updated as per the approach discussed above. Boards converted to variants.

**Comment by jacobw** [comment]
Ping @anangl @decsny  as per PR Review Escalation. It's been a week since all checks passed.

**Comment by decsny** [comment]
im not actually sure why this got assigned to the ADC area

**Comment by decsny** [comment]
added @nordicjm as he has been reviewing hwmv2 board rework type stuff

File: boards/adafruit/feather_nrf52840/Kconfig [hunk 1]
```diff
@@ -1,12 +1,12 @@
-# Adafruit Feather nRF52840 Express board configuration
+# Adafruit Feather nRF52840 boards configuration
 
 # Copyright (c) 2020 Tobias Svehagen
 # Copyright (c) 2024 Jacob Winther
 # SPDX-License-Identifier: Apache-2.0
 
-if BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE
+if (BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE || BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)
 config BOARD_SERIAL_BACKEND_CDC_ACM
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/Kconfig]:
```suggestion
if(BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE || BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)

config BOARD_SERIAL_BACKEND_CDC_ACM
```

File: boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2.dts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+		/* To enable flashing with UF2 bootloader, we
+		 * must reserve a partition for SoftDevice.
+		 * See https://learn.adafruit.com/
+		 *     introducing-the-adafruit-nrf52840-feather?view=all#hathach-memory-map
+		 */
+		reserved_partition_0: partition@0 {
+			label = "SoftDevice";
+			reg = <0x000000000 DT_SIZE_K(152)>;
+		};
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2.dts]: would add `read-only;`

File: boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2.dts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		 */
+		storage_partition: partition@ed000 {
+			label = "storage";
+			reg = <0x0000ed000 DT_SIZE_K(28)>;
+		};
+
+		boot_partition: partition@f4000 {
+			label = "UF2";
+			reg = <0x000f4000 DT_SIZE_K(48)>;
+		};
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2.dts]: likewise here too

File: boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# enable GPIO
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig]:
```suggestion
# Enable GPIO
```

File: boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+CONFIG_GPIO=y
+
+# enable uart driver
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig]:
```suggestion
# Enable UART driver
```

File: boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+CONFIG_SERIAL=y
+
+# enable console
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig]:
```suggestion
# Enable console
```

File: boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# enable bossac
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig]:
```suggestion
# Enable bossac
```

File: boards/adafruit/feather_nrf52840/board.cmake [hunk 8]
```diff
@@ -2,6 +2,9 @@
 
 board_runner_args(jlink "--device=nRF52840_xxAA" "--speed=4000")
 board_runner_args(pyocd "--target=nrf52840" "--frequency=4000000")
+if(CONFIG_BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)
+include(${ZEPHYR_BASE}/boards/common/uf2.board.cmake)
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/board.cmake]: 2 space indent for cmake

File: boards/adafruit/feather_nrf52840/board.cmake [hunk 9]
```diff
@@ -2,6 +2,9 @@
...
+endif()
 if(CONFIG_BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE)
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/board.cmake]: elseif()

File: boards/adafruit/feather_nrf52840/doc/index.rst [hunk 10]
```diff
@@ -142,6 +142,10 @@ The Feather nRF52840 ships with the `Adafruit nRF52 Bootloader`_ which
 supports flashing using `UF2`_. This allows easy flashing of new images,
 but does not support debugging the device.
 
+To flash the ``adafruit_feather_nrf52480_express`` using UF2, you must use
+the ``adafruit_feather_nrf52480_express_uf2`` board, which uses an appropriate
+flash layout.
```
Comment (nordicjm) [inline boards/adafruit/feather_nrf52840/doc/index.rst]: I don't get it, if this is actually the same board, then this should be a variant, not a new board

**Comment by nordicjm** [comment]
> Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.

Having see this just now, yes this is a wrong design, it needs to be a variant of the existing board

**Comment by jacobw** [comment]
> > Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.
> 
> Having see this just now, yes this is a wrong design, it needs to be a variant of the existing board

Thanks for the review. It took my a long time to realise that BOARD_$(BOARD_TARGET_STRING) was an option, but I work around that and converted everything to variants. Is that what you mean. I see I missed that in the documentation so will fix,

**Comment by nordicjm** [comment]
> > > Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.
> > 
> > 
> > Having see this just now, yes this is a wrong design, it needs to be a variant of the existing board
> 
> Thanks for the review. It took my a long time to realise that BOARD_$(BOARD_TARGET_STRING) was an option, but I work around that and converted everything to variants. Is that what you mean. I see I missed that in the documentation so will fix,

Best way to see it would be an existing board, so https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/nordic/nrf54l15pdk is an example, if you check the board.yml file you will see:
```
board:
  name: nrf54l15pdk
  vendor: nordic
  socs:
  - name: nrf54l15
    variants:
    - name: xip
      cpucluster: cpuflpr
```

This means that there is a board nrf54l15pdk, it has a soc nrf54l15, this soc has multiple CPU cores, for the cpuflpr core, you can optionally build just for that or you can build for the XIP variant of that, if you build for the XIP variant, it uses the other configuration

**Comment by jacobw** [comment]
> Best way to see it would be an existing board

I convert to variants in a later commit in this PR. Sorry if that was confusing, I covert the existing boards to variants in the same commit.

**Comment by jacobw** [comment]
@nordicjm I realise I've structured this PR poorly. I'll restructure it and submit it as a new PR. Apologies for wasting your time.

**Pull request closed without merging**